cryptoki/examples/thread_local_session.rs

```diff
@@ -0,0 +1,294 @@
+//! # Thread-Local Session Pattern Example
+//!
+//! This example demonstrates how to safely use cryptoki in a multi-threaded
+//! application by combining:
+//!
+//! 1. **Shared Pkcs11 context** - One context for all threads (via Arc)
+//! 2. **Thread-local Sessions** - Each thread has its own Session (via thread_local!)
+//!
+//! ## Why This Pattern?
+//!
+//! - **Pkcs11 context**: Is Send + Sync, can be shared via Arc (cheap clone)
+//! - **Session**: Is Send but NOT Sync (can transfer ownership, cannot share)
+//!
+//! The Session type deliberately prevents sharing across threads by not
+//! implementing Sync. This matches PKCS#11 C specification where sessions
+//! are not thread-safe.
+//!
+//! ## Architecture
+//!
+//! ```text
+//! +---------------------------------------+
+//! | static PKCS11_CTX: Arc<Pkcs11>        | <- Shared across threads
+//! +---------------------------------------+
+//!          |
+//!          +---> Thread 1: thread_local! { Session<'_> }
+//!          +---> Thread 2: thread_local! { Session<'_> }
+//!          +---> Thread 3: thread_local! { Session<'_> }
+//! ```
+//!
+//! ## Running This Example
+//!
+//! ```bash
+//! export TEST_PKCS11_MODULE=/usr/local/lib/softhsm/libsofthsm2.so
+//! cargo run --example thread_local_session
+//! ```
+
+use std::cell::RefCell;
+use std::env;
+use std::sync::{Arc, OnceLock};
+use std::thread;
+
+use cryptoki::context::{CInitializeArgs, CInitializeFlags, Pkcs11};
+use cryptoki::mechanism::Mechanism;
+use cryptoki::object::Attribute;
+use cryptoki::session::{Session, UserType};
+use cryptoki::types::AuthPin;
+
+const USER_PIN: &str = "fedcba";
+const SO_PIN: &str = "abcdef";
+
+// Global PKCS11 context shared across all threads using Arc for cheap cloning
+static PKCS11_CTX: OnceLock<Arc<Pkcs11>> = OnceLock::new();
```

Comment:
We moved back with `Pkcs11` containing an `Arc` so maybe it's not needed to have another Arc here :)